File: server/python/src/metaobjects/super_resolve.py
"""Deferred super/extends resolution over the merged tree (2nd pass, pre-freeze).

Mirrors TS ``resolveDeferredSupers`` in super-resolve.ts:
- Build the FQN index keyed by node.fqn() (own), over the whole merged tree.
- Collect every node carrying an unresolved super_ref, tracking each node's
  inherited context package.
- Resolve supers ON DEMAND with memoization + cycle detection (#188), NOT in a
  single physical-declaration-order pass — so resolution is order-independent.
"""
from __future__ import annotations

from .errors import ErrorCode, MetaError
from .meta.meta_data import MetaData
from .shared.separators import PACKAGE_SEP
from .source import resolved_source


def resolve_supers(root: MetaData, errors: list[MetaError]) -> None:
    """Resolve every unresolved super_ref in the merged tree, order-independently.

    Resolved → sets node.super_data.
    Unresolved → appends ERR_UNRESOLVED_SUPER to errors.
    Dotted target type/subtype mismatch → appends ERR_EXTENDS_TARGET_MISMATCH.
    Already-resolved nodes (super_data is not None) are skipped (idempotent).

    #188: super-resolution must be ORDER-INDEPENDENT. A pre-order walk resolved
    each node's super_ref in physical-declaration (= load) order, so a dotted ref
    to an INHERITED member (``extends: Owner.member`` where Owner inherits
    ``member`` via its OWN ``extends``) only succeeded when the owner's extends
    chain happened to be wired first — green under one readdir order, throwing
    ERR_UNRESOLVED_SUPER under another. Instead resolve ON DEMAND with
    memoization + cycle detection: before a dotted ref reads ``owner.children()``
    (EFFECTIVE children — inherited members appear only once the owner's OWN
    super is resolved), resolve the owner's whole chain first. The result is a
    pure function of the source SET, independent of enumeration order. Mirrors
    the TS reference (#188).
    """
    index = _build_index(root)

    # Every node carrying a super_ref, over the PHYSICAL declaration tree, with
    # the inherited context package captured at that node's walk position (the
    # effective_pkg the original pre-order walk computed — kept byte-identical so
    # every existing fixture resolves the same way).
    pending: list[MetaData] = []
    pkg_of: dict[MetaData, str | None] = {}
    _collect(root, "", pending, pkg_of)

    # Nodes already attempted this pass. On-demand resolution can reach a node
    # via the ``pending`` loop AND via owner/target recursion — ``attempted``
    # makes each node resolve (and, on failure, report) EXACTLY ONCE, restoring
    # the single-visit walk's no-duplicate-failures guarantee (a successful node
    # is deduped by ``super_data``; this also covers the FAILURE path, which
    # sets no marker). It is also the cycle guard: ``attempted`` is never
    # removed, so a genuine super cycle (A→B→A) terminates on re-entry to the
    # already-attempted node rather than recursing forever. Mirrors the TS
    # reference (#188).
    attempted: set[MetaData] = set()

    def resolve_node(node: MetaData) -> None:
        if not node.super_ref or node.super_data is not None:
            return
        if node in attempted:
            return  # already resolved-or-failed this pass — never re-report
        attempted.add(node)
        # Referrer context package: own ``package`` if declared, else the
        # file-default package captured at parse time, else the inherited walk
        # context. Captured during _collect (mirrors the original ``_walk``
        # effective_pkg). Using file_default_package (not just the threaded
        # ctx_pkg) is what lets a bare / same-package / cross-package ``extends``
        # resolve over the MERGED tree.
        effective_pkg = pkg_of[node] if node in pkg_of else (
            node.package or node.file_default_package or None
        )

        # For a dotted child-targeting ref, ``_resolve`` reads the OWNER's
        # effective ``children()`` — which only includes inherited members once
        # the owner's own extends chain is resolved. So resolve the owner node's
        # chain FIRST (memoized), regardless of load order (#188).
        if _is_child_targeting_ref(node.super_ref):
            owner_ref = _owner_ref(node.super_ref)
            if owner_ref is not None:
                owner = _resolve(owner_ref, effective_pkg, index)
                if owner is not None:
                    resolve_node(owner)

        # FR-024: thread the referrer's type so dotted ``Entity.child`` refs
        # resolve type-scoped (a field ref selects fields; an identity ref
        # identities).
        target = _resolve(node.super_ref, effective_pkg, index, referrer_type=node.type)

        if target is None:
            # FR5d / ADR-0009: emit a ResolvedSource envelope carrying the
            # referrer's files / json_path plus the referrer FQN + unresolved
            # target. Mirrors TS resolveDeferredSupers in meta-data-loader.ts.
            errors.append(MetaError(
                f"the SuperClass '{node.super_ref}' does not exist "
                f"(referenced by {node.fqn()})",
                ErrorCode.ERR_UNRESOLVED_SUPER,
                path=node.fqn(),
                envelope=resolved_source(node.source, node.fqn(), node.super_ref),
            ))
        elif _is_child_targeting_ref(node.super_ref) and (
            target.type != node.type or target.sub_type != node.sub_type
        ):
            # FR-024 (ADR-0029): a dotted extends target must match the referrer's
            # type AND subtype (dotted-only check — top-level extends unchanged).
            # Mirrors TS resolveDeferredSupers' target-mismatch kind.
            errors.append(MetaError(
                f"the dotted extends target '{node.super_ref}' is "
                f"[{target.type}.{target.sub_type}] but the extending node is "
                f"[{node.type}.{node.sub_type}] — a dotted extends must target a "
                f"node of the same type and subtype (referenced by {node.fqn()})",
                ErrorCode.ERR_EXTENDS_TARGET_MISMATCH,
                path=node.fqn(),
                envelope=resolved_source(node.source, node.fqn(), node.super_ref),
            ))
        else:
            node.super_data = target
            # Ensure the target's OWN chain is resolved too, so a later
            # ``children()`` read on *node* (codegen/validation) sees the full
            # multi-level inherited set (e.g. Base extends AbstractRoot).
            # Memoized — no re-work.
            resolve_node(target)

    for node in pending:
        resolve_node(node)


def _collect(
    node: MetaData,
    ctx_pkg: str,
    pending: list[MetaData],
    pkg_of: dict[MetaData, str | None],
) -> None:
    """Pre-order walk: record every node carrying a super_ref plus the effective
    context package at its walk position, then recurse over own_children().

    Mirrors the referrer-context computation of the original ``_walk`` so the
    on-demand resolver uses exactly the same effective_pkg the pre-order walk
    did — resolution stays byte-identical for every existing fixture; only the
    ORDER in which supers are wired changes (now order-independent, #188).
    """
    if node.super_ref:
        # own ``package`` if declared, else the file-default package captured at
        # parse time, else the inherited walk context.
        pkg_of[node] = node.package or node.file_default_package or ctx_pkg or None
        pending.append(node)

    # The context package for children is this node's own package, if set, else inherit.
    next_ctx = node.package or ctx_pkg
    # ADR-0039 sanctioned own: this IS the super-resolution collection walk — it
    # runs BEFORE extends is resolved (resolution sets super_data), so it must
    # iterate own children.
    for child in node.own_children():
        _collect(child, next_ctx, pending, pkg_of)


def _build_index(root: MetaData) -> dict[str, MetaData]:
    """Build a lookup index over the whole merged tree (own_children walk).

    Each named node is registered under its own ``fqn()`` AND — when it has no
    own ``package`` but a file-default package was captured at parse time —
    under the package-folded key ``<file_default_package>::<name>``. The second
    key is what makes a cross-package fully-qualified ``extends`` resolve over
    the merged tree (object ``fqn()`` stays bare because the parser does not
    fold the file-default package onto the object's own package). Mirrors the
    TS ``findInTree`` matcher (own ``fqn()`` OR ``resolutionKey()``).
    """
    idx: dict[str, MetaData] = {}
    _index_walk(root, idx)
    return idx


def _index_walk(node: MetaData, idx: dict[str, MetaData]) -> None:
    if node.name:
        idx.setdefault(node.fqn(), node)
        if not node.package and node.file_default_package:
            folded = f"{node.file_default_package}{PACKAGE_SEP}{node.name}"
            idx.setdefault(folded, node)
    # ADR-0039 sanctioned own: index-building walk over the declared tree (feeds
    # super-resolution; runs before extends is resolved).
    for child in node.own_children():
        _index_walk(child, idx)


def _is_child_targeting_ref(ref: str) -> bool:
    """FR-024 (ADR-0029): True when the ref's final ``::``-segment contains a
    ``.`` — i.e. it targets a child nested inside an object (``Customer.id``,
    ``acme::sales::Customer.id``). Names cannot contain ``.``, so the form is
    unambiguous."""
    last_sep = ref.rfind(PACKAGE_SEP)
    last_segment = ref if last_sep < 0 else ref[last_sep + len(PACKAGE_SEP):]
    return "." in last_segment


def _owner_ref(ref: str) -> str | None:
    """The OWNER portion of a dotted child-targeting ref (``Owner.child`` →
    ``Owner``, ``pkg::Owner.child.grandchild`` → ``pkg::Owner``): the package
    prefix plus the first dotted segment. ``None`` for a degenerate ref (fewer
    than two segments or any empty segment). Mirrors the TS
    ``parseChildTargetingRef().ownerRef``."""
    last_sep = ref.rfind(PACKAGE_SEP)
    seg_start = 0 if last_sep < 0 else last_sep + len(PACKAGE_SEP)
    parts = ref[seg_start:].split(".")
    if len(parts) < 2 or any(not p for p in parts):
        return None
    return ref[:seg_start] + parts[0]


def _resolve(
    ref: str,
    context_pkg: str | None,
    index: dict[str, MetaData],
    referrer_type: str | None = None,
) -> MetaData | None:
    """Resolve a super_ref string against the FQN index.

    Resolution forms:
    - absolute ``::pkg::Name``  → strip leading ``::``, look up ``pkg::Name``.
    - relative ``..::rest``     → count leading ``..::`` levels; if levels exceed
                                  context depth or remainder is empty → ``None``
                                  (→ ERR_UNRESOLVED_SUPER); else look up
                                  ``reducedCtx::rest`` (mirrors TS exactly).
    - bare/qualified ``Name``   → try ``context::ref`` first, then bare ``ref``.
    - dotted ``Owner.child``    → FR-024 (ADR-0029): resolve the OWNER with the
                                  strategies above, then select the child among
                                  the owner's EFFECTIVE children (``children()``,
                                  so inherited children resolve) by name + the
                                  REFERRER's type (type-scoped: a field ref
                                  resolves fields, an identity ref identities).
                                  Dotted refs never fall through to the bare
                                  lookup; multi-dot (``X.y.z``) TRAVERSES child
                                  names to any depth (intermediate segments by
                                  unique name); only degenerate empty segments →
                                  ``None``; no ``referrer_type`` → ``None``.
                                  Mirrors TS super-resolve.ts.
    """
    abs_prefix = PACKAGE_SEP  # "::"
    rel_prefix = ".." + PACKAGE_SEP  # "..::

    if _is_child_targeting_ref(ref):
        # Addressing model (ADR-0029): the package qualifies the ROOT-level node
        # only; each subsequent segment traverses CHILD NAMES to any depth
        # (object → field → view: ``Customer.priceCents.display``). INTERMEDIATE
        # segments select by UNIQUE name among the current node's effective
        # children (a cross-type name collision is ambiguous → unresolved); the
        # FINAL segment is type-scoped to the referrer. Mirrors the TS reference.
        if referrer_type is None:
            return None
        last_sep = ref.rfind(PACKAGE_SEP)
        seg_start = 0 if last_sep < 0 else last_sep + len(PACKAGE_SEP)
        parts = ref[seg_start:].split(".")
        if len(parts) < 2 or any(not p for p in parts):
            return None  # degenerate (empty segment)
        current = _resolve(ref[:seg_start] + parts[0], context_pkg, index)
        if current is None:
            return None
        for seg in parts[1:-1]:
            matches = [c for c in current.children() if c.name == seg]
            if len(matches) != 1:
                return None  # missing or ambiguous intermediate
            current = matches[0]
        last = parts[-1]
        for child in current.children():
            if child.name == last and child.type == referrer_type:
                return child
        return None

    if ref.startswith(abs_prefix):                          # absolute ::pkg::Name
        return index.get(ref[len(abs_prefix):])

    if ref.startswith(rel_prefix):                          # relative ..::rest
        parts = ref.split(PACKAGE_SEP)
        levels = 0
        while levels < len(parts) and parts[levels] == "..":
            levels += 1
        pkg_parts = context_pkg.split(PACKAGE_SEP) if context_pkg else []
        remainder = parts[levels:]
        if len(pkg_parts) < levels or len(remainder) == 0:
            return None
        all_parts = pkg_parts[: len(pkg_parts) - levels] + remainder
        return index.get(PACKAGE_SEP.join(all_parts))

    # bare or pkg-qualified (no leading :: / ..)
    if context_pkg:
        hit = index.get(f"{context_pkg}{PACKAGE_SEP}{ref}")
        if hit is not None:
            return hit
    return index.get(ref)

File: server/csharp/MetaObjects/SuperResolve.cs
// super reference resolution — package navigation helpers
//
// Ported from typescript/packages/metadata/src/super-resolve.ts.
//
// Reference syntax (Java metaobjects-core conventions):
//   Bare name       "Fruit"            → resolve in contextPackage, then root fallback
//   Absolute        "::fishstore::Fish" → strip leading ::, walk from root
//   Relative parent "..::common::id"   → up one package level, then resolve "common::id"
//   Multi-level     "..::..::shared::User" → up two levels, then resolve "shared::User"
//   Same-package    "common::id"       → try contextPackage prepend first, then root-rooted
//
// Super resolution is IMMEDIATE during parse (Java behavior). This module
// provides only the lookup helper; the parser calls it inline. The deferred
// multi-pass walker (resolveDeferredSupers) is added in Task 5.1.

using MetaObjects.Meta;

namespace MetaObjects;

/// <summary>
/// Super reference resolution helpers — package navigation against a tree.
/// </summary>
internal static class SuperResolve
{
    // -----------------------------------------------------------------------
    // Tree search helper
    // -----------------------------------------------------------------------

    /// <summary>
    /// Recursively searches the tree for a node whose <see cref="MetaData.Fqn"/>
    /// matches. Returns the first match, or null.
    /// </summary>
    private static MetaData? FindInTree(MetaData root, string fqn)
    {
        if (root.Fqn() == fqn) return root;
        // Match by EFFECTIVE qualified key (file-default package folded onto the
        // bare name, captured at parse time) for nodes carrying no own package —
        // the cross-PACKAGE cross-file case. Mirrors the TS findInTree second
        // clause. The parser keeps an object's Fqn() bare for the FR5d
        // referrer-envelope contract, so this clause is what lets a fully-qualified
        // `extends: acme::common::BaseEntity` resolve against a `BaseEntity`
        // declared under `package: acme::common` in a separate file.
        if (root.Package is null && root.Name != "" && root.ResolutionKey() == fqn)
        {
            return root;
        }
        // ADR-0039: OwnChildren — super-resolution tree walk. This IS the extends-resolution
        // machinery, so it must traverse only the physically-declared children (resolving
        // Children() would recurse through the very super chain being built).
        foreach (MetaData child in root.OwnChildren())
        {
            MetaData? found = FindInTree(child, fqn);
            if (found is not null) return found;
        }
        return null;
    }

    // -----------------------------------------------------------------------
    // Reference parsing & resolution
    // -----------------------------------------------------------------------

    /// <summary>
    /// FR-024 (ADR-0029): the type-scope of the node whose <c>extends</c> is being
    /// resolved. A dotted <c>Entity.child</c> ref selects among the owner's children of
    /// the SAME type as the referrer — a field ref resolves fields, an identity
    /// ref resolves identities. Required for the dotted branch; ignored for
    /// dot-free refs (legacy behavior unchanged).
    /// </summary>
    public sealed record ReferrerScope(string Type);

    /// <summary>
    /// FR-024: true when a ref's final <c>::</c>-segment contains a <c>.</c> — i.e. the ref
    /// targets a child nested inside an object (<c>Customer.id</c>,
    /// <c>acme::sales::Customer.id</c>). Names cannot contain <c>.</c>, so this is
    /// unambiguous. Call sites use this to apply the dotted-only
    /// type/subtype-mismatch check (ERR_EXTENDS_TARGET_MISMATCH) without altering
    /// shipped top-level extends behavior.
    /// </summary>
    public static bool IsChildTargetingRef(string reference)
    {
        int lastSep = reference.LastIndexOf(PACKAGE_SEPARATOR, StringComparison.Ordinal);
        string lastSegment = lastSep == -1 ? reference : reference[(lastSep + PACKAGE_SEPARATOR.Length)..];
        return lastSegment.Contains(CHILD_REF_SEPARATOR, StringComparison.Ordinal);
    }

    /// <summary>
    /// FR-024: split a child-targeting ref into the root-object ref and the child
    /// traversal path. The addressing model: a package qualifies the ROOT-level
    /// node only; every subsequent dotted segment traverses CHILD NAMES to any
    /// depth (<c>Customer.id</c>, <c>Customer.priceCents.display</c> — object →
    /// field → view). Returns null for degenerate empty parts (<c>.id</c>,
    /// <c>Customer.</c>, <c>Customer..x</c>).
    /// </summary>
    private static (string OwnerRef, string[] Path)? ParseChildTargetingRef(string reference)
    {
        int lastSep = reference.LastIndexOf(PACKAGE_SEPARATOR, StringComparison.Ordinal);
        int segStart = lastSep == -1 ? 0 : lastSep + PACKAGE_SEPARATOR.Length;
        string lastSegment = reference[segStart..];
        string[] parts = lastSegment.Split(CHILD_REF_SEPARATOR, StringSplitOptions.None);
        if (parts.Length < 2 || Array.Exists(parts, p => p == "")) return null;
        return (reference[..segStart] + parts[0], parts[1..]);
    }

    /// <summary>
    /// Resolve a single super reference string against a tree.
    ///
    /// <para>
    /// FR-024 (ADR-0029): a ref whose final segment is dotted (<c>Customer.id</c>)
    /// targets a child nested inside an object. The owner part resolves with the
    /// existing strategies (absolute / relative / bare / same-package); the child
    /// is then selected among the owner's EFFECTIVE children (so inherited
    /// children resolve) by name AND the referrer's type (type-scoped). A dotted
    /// ref that does not resolve returns null WITHOUT falling through to the
    /// bare lookup; a dotted ref with no <paramref name="referrerScope"/> is unresolvable.
    /// </para>
    /// </summary>
    /// <param name="reference">
    /// The raw super reference (e.g. "Fruit", "::pkg::Name", "..::common::id", "Customer.id").
    /// </param>
    /// <param name="contextPackage">
    /// The package of the model whose super is being resolved.
    /// </param>
    /// <param name="root">The root MetaData of the accumulating tree to search within.</param>
    /// <param name="referrerScope">FR-024: the extending node's type — required to resolve dotted refs.</param>
    /// <returns>The resolved MetaData, or null if the reference cannot be resolved.</returns>
    public static MetaData? ResolveSuperRef(
        string reference,
        string contextPackage,
        MetaData root,
        ReferrerScope? referrerScope = null)
    {
        // ---------------------------------------------------------------------
        // 0. FR-024 dotted child-targeting ref: `<rootRef>.<child>...<child>`
        //
        // Addressing model (ADR-0029): the package qualifies the ROOT-level node
        // only; each subsequent segment traverses CHILD NAMES to any depth
        // (object → field → view: `Customer.priceCents.display`). INTERMEDIATE
        // segments select by UNIQUE name among the current node's effective
        // children (a cross-type name collision is ambiguous → unresolved); the
        // FINAL segment is type-scoped to the referrer. Mirrors the TS reference.
        // ---------------------------------------------------------------------
        if (IsChildTargetingRef(reference))
        {
            if (referrerScope is null) return null;
            (string OwnerRef, string[] Path)? parsed = ParseChildTargetingRef(reference);
            if (parsed is null) return null; // degenerate (empty segment)
            MetaData? current = ResolveSuperRef(parsed.Value.OwnerRef, contextPackage, root);
            if (current is null) return null;
            for (int i = 0; i < parsed.Value.Path.Length - 1; i++)
            {
                string seg = parsed.Value.Path[i];
                var matches = current.Children().Where(c => c.Name == seg).ToList();
                if (matches.Count != 1) return null; // missing or ambiguous intermediate
                current = matches[0];
            }
            string last = parsed.Value.Path[^1];
            return current
                .Children()
                .FirstOrDefault(c => c.Name == last && c.Type == referrerScope.Type);
        }

        // ---------------------------------------------------------------------
        // 1. Absolute reference: leading "::"
        // ---------------------------------------------------------------------
        if (reference.StartsWith(PACKAGE_SEPARATOR, StringComparison.Ordinal))
        {
            string absolutePath = reference[PACKAGE_SEPARATOR.Length..];
            return FindInTree(root, absolutePath);
        }

        // ---------------------------------------------------------------------
        // 2. Relative reference: one or more leading "..::"
        // ---------------------------------------------------------------------
        if (reference.StartsWith(PACKAGE_PARENT + PACKAGE_SEPARATOR,
                StringComparison.Ordinal))
        {
            string[] parts = Split(reference);
            int levels = 0;
            while (levels < parts.Length && parts[levels] == PACKAGE_PARENT)
            {
                levels++;
            }

            string[] pkgParts = contextPackage != "" ? Split(contextPackage) : [];
            string[] remainder = parts.Skip(levels).ToArray();
            if (pkgParts.Length < levels || remainder.Length == 0)
            {
                return null;
            }

            IEnumerable<string> allParts = pkgParts.Take(pkgParts.Length - levels).Concat(remainder);
            return FindInTree(root, string.Join(PACKAGE_SEPARATOR, allParts));
        }

        // ---------------------------------------------------------------------
        // 3. Bare name OR same-package shorthand (contains "::" but no leading
        //    "::" or "..::"): try contextPackage prepend first, then root-rooted.
        // ---------------------------------------------------------------------
        if (contextPackage != "")
        {
            MetaData? found = FindInTree(root,
                $"{contextPackage}{PACKAGE_SEPARATOR}{reference}");
            if (found is not null) return found;
        }
        return FindInTree(root, reference);
    }

    /// <summary>
    /// Split a package path on the <c>::</c> separator. Mirrors the JS
    /// <c>String.prototype.split</c> on a multi-char separator.
    /// </summary>
    private static string[] Split(string value) =>
        value.Split(PACKAGE_SEPARATOR, StringSplitOptions.None);

    // -----------------------------------------------------------------------
    // Deferred resolution — second pass after all files parsed
    // -----------------------------------------------------------------------

    /// <summary>
    /// FR-024: why deferred super resolution failed.
    /// <list type="bullet">
    /// <item><see cref="Unresolved"/> — no target found (loader emits ERR_UNRESOLVED_SUPER).</item>
    /// <item><see cref="TargetMismatch"/> — a dotted child-targeting ref resolved, but the
    /// target's type/subtype differs from the extending node's (loader emits
    /// ERR_EXTENDS_TARGET_MISMATCH). Applies ONLY to dotted refs — top-level
    /// extends behavior is unchanged.</item>
    /// </list>
    /// </summary>
    public enum SuperFailureKind
    {
        Unresolved,
        TargetMismatch,
    }

    /// <summary>A node's type + subType identity, for the target-mismatch message.</summary>
    public sealed record TypeIdentity(string Type, string SubType);

    /// <summary>One node whose extends reference could not be resolved against the full tree.
    /// <see cref="Node"/> is the offending referrer node (useful for FR5a / ADR-0009 to
    /// attach <c>node.Source</c> to the loader error envelope).
    /// FR-024: <see cref="Kind"/> distinguishes unresolved refs from dotted target-mismatch;
    /// <see cref="Target"/> / <see cref="Referrer"/> are populated for target-mismatch only.</summary>
    public sealed record DeferredSuperFailure(
        string NodeFqn,
        string Ref,
        MetaData? Node = null,
        SuperFailureKind Kind = SuperFailureKind.Unresolved,
        TypeIdentity? Target = null,
        TypeIdentity? Referrer = null);

    /// <summary>
    /// Resolve every node's SuperRef against the full root, collecting unresolved
    /// refs. Idempotent: nodes that already have SuperData set are skipped.
    ///
    /// <para>
    /// #188: super-resolution is ORDER-INDEPENDENT. The prior implementation
    /// resolved each SuperRef in a single pre-order walk — physical (= load)
    /// order. A dotted ref to an INHERITED member (<c>extends: Owner.member</c>
    /// where Owner inherits <c>member</c> via its OWN <c>extends</c>) reads the
    /// owner's EFFECTIVE <see cref="MetaData.Children"/> — inherited members appear
    /// only once the owner's extends chain is resolved. So it succeeded only when
    /// the owner happened to resolve first: green under one directory-scan order,
    /// ERR_UNRESOLVED_SUPER under another. Instead resolve ON DEMAND with
    /// memoization + cycle detection (topological): before a dotted ref reads the
    /// owner's children, resolve the owner's whole chain first, and after resolving
    /// a node also resolve its target's chain (multi-level inheritance). The result
    /// is a pure function of the source SET, independent of enumeration order.
    /// </para>
    ///
    /// <para>
    /// An inherited context-package is threaded while collecting the pending set so
    /// children whose own Package is unset still resolve correctly (mirroring the
    /// parser's eager-resolution semantics); it is captured per node so a node
    /// resolved out of collection order still uses the correct package.
    /// </para>
    /// </summary>
    public static IReadOnlyList<DeferredSuperFailure> ResolveDeferredSupers(MetaData root)
    {
        var failures = new List<DeferredSuperFailure>();

        // Every SuperRef-bearing node over the PHYSICAL declaration tree, plus each
        // node's effective context-package (node.Package ?? nearest-ancestor
        // package) — the same value the pre-order walk threaded.
        var pending = new List<MetaData>();
        var effectivePkgOf = new Dictionary<MetaData, string>(ReferenceEqualityComparer.Instance);
        Walk(root, "", (node, ctxPkg) =>
        {
            if (node.SuperRef is null) return;
            effectivePkgOf[node] = node.Package ?? ctxPkg;
            pending.Add(node);
        });

        // Cycle guard: a genuine super cycle is unresolvable → the node is left
        // unresolved and reported as a failure below.
        var inProgress = new HashSet<MetaData>(ReferenceEqualityComparer.Instance);
        // Nodes already attempted this pass. On-demand resolution can reach a node
        // via the `pending` loop AND via owner/target recursion — `attempted` makes
        // each node resolve (and, on failure, report) EXACTLY ONCE, restoring the
        // single-visit walk's no-duplicate-failures guarantee (a successful node is
        // deduped by SuperData; this also covers the FAILURE path, which sets no
        // marker). Mirrors the TS reference (#188).
        var attempted = new HashSet<MetaData>(ReferenceEqualityComparer.Instance);

        void ResolveNode(MetaData node)
        {
            if (node.SuperRef is null || node.SuperData is not null) return;
            if (!attempted.Add(node)) return; // already resolved-or-failed this pass — never re-report
            if (!inProgress.Add(node)) return; // cycle — leave unresolved; reported below
            string effectivePkg = effectivePkgOf.TryGetValue(node, out var pkg) ? pkg : (node.Package ?? "");

            // For a dotted child-targeting ref, ResolveSuperRef reads the OWNER's
            // effective Children() — which includes inherited members only once the
            // owner's own extends chain is resolved. Resolve the owner node's chain
            // FIRST (memoized), regardless of collection order.
            if (IsChildTargetingRef(node.SuperRef))
            {
                (string OwnerRef, string[] Path)? parsed = ParseChildTargetingRef(node.SuperRef);
                if (parsed is not null)
                {
                    MetaData? ownerNode = ResolveSuperRef(parsed.Value.OwnerRef, effectivePkg, root);
                    if (ownerNode is not null) ResolveNode(ownerNode);
                }
            }

            // FR-024: thread the referrer's type so dotted `Entity.child` refs resolve
            // type-scoped (a field ref selects fields; an identity ref identities).
            MetaData? target = ResolveSuperRef(node.SuperRef, effectivePkg, root, new ReferrerScope(node.Type));
            inProgress.Remove(node);

            if (target is not null)
            {
                // FR-024: a dotted ref must target a node of the SAME type and subtype
                // as the extending node. Dotted-only — top-level extends is unchanged.
                if (IsChildTargetingRef(node.SuperRef) &&
                    (target.Type != node.Type || target.SubType != node.SubType))
                {
                    failures.Add(new DeferredSuperFailure(
                        node.Fqn(),
                        node.SuperRef,
                        node,
                        SuperFailureKind.TargetMismatch,
                        new TypeIdentity(target.Type, target.SubType),
                        new TypeIdentity(node.Type, node.SubType)));
                    return;
                }
                try
                {
                    node.SetSuperResolved(target);
                }
                catch (InvalidOperationException)
                {
                    // Frozen — ignore; the loader should resolve before freeze.
                }
                // Ensure the target's OWN chain is resolved too, so a later Children()
                // read on `node` (codegen/validation) sees the full multi-level
                // inherited set (e.g. Base extends AbstractRoot). Memoized — no re-work.
                ResolveNode(target);
            }
            else
            {
                failures.Add(new DeferredSuperFailure(node.Fqn(), node.SuperRef, node));
            }
        }

        foreach (var node in pending) ResolveNode(node);
        return failures.AsReadOnly();
    }

    private static void Walk(MetaData node, string ctxPkg, Action<MetaData, string> visit)
    {
        visit(node, ctxPkg);
        string nextCtx = node.Package ?? ctxPkg;
        // ADR-0039: OwnChildren — this is a declaration-structure walk internal to
        // super-resolution; it must visit only physically-declared children.
        foreach (MetaData child in node.OwnChildren())
        {
            Walk(child, nextCtx, visit);
        }
    }
}
